cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java

```diff
@@ -0,0 +1,131 @@
+ */
+public class DefaultPreviewStore implements PreviewStore {
+
+  private static final DatasetId PREVIEW_TABLE_ID = NamespaceId.SYSTEM.dataset(PreviewDataset.PREVIEW_TABLE_NAME);
+
+  private final DatasetFramework dsFramework;
+  private final Transactional transactional;
+
+  @Inject
+  public DefaultPreviewStore(final TransactionExecutorFactory txExecutorFactory, DatasetFramework framework,
+                             TransactionSystemClient txClient) {
+
+    this.dsFramework = framework;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(framework), txClient,
+        NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+  }
+
+  @Override
+  public void put(final ApplicationId applicationId, final String loggerName, final String propertyName,
+                  final Object value) {
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          getPreviewDataset(context).put(applicationId, loggerName, propertyName, value);
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e);
+    }
+  }
+
+  @Override
+  public Map<String, List<String>> get(final ApplicationId applicationId, final String loggerName) {
+    try {
+      return Transactions.execute(transactional, new TxCallable<Map<String, List<String>>>() {
+        @Override
+        public Map<String, List<String>> call(DatasetContext context) throws Exception {
+          return getPreviewDataset(context).get(applicationId, loggerName);
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e);
+    }
+  }
+
+  @Override
+  public void remove(final ApplicationId applicationId) {
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          getPreviewDataset(context).remove(applicationId);
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e);
+    }
+  }
+
+  @VisibleForTesting
+  @Override
+  public void clear() throws IOException, DatasetManagementException {
+    truncate(dsFramework.getAdmin(PREVIEW_TABLE_ID, null));
+  }
+
+  private PreviewDataset getPreviewDataset (DatasetContext datasetContext)
+    throws IOException, DatasetManagementException {
+    Table table =  DatasetsUtil.getOrCreateDataset(datasetContext, dsFramework, PREVIEW_TABLE_ID, Table.class.getName(),
+                                                   DatasetProperties.EMPTY);
+    return new PreviewDataset(table);
+  }
+
+  private void truncate(DatasetAdmin admin) throws IOException {
+    if (admin != null) {
```

Comment:
The admin is get from `dsFrameWork.getAdmin(name)`, which will be null if dataset instances of the provided name does not exist.